xlators/mgmt/glusterd/src/glusterd-pmap.c

```diff
@@ -152,275 +188,187 @@ pmap_registry_search(xlator_t *this, const char *brickname,
     char *brick_str;
 
-    priv = this->private;
-    pmap = priv->pmap;
-    if (!pmap)
-        goto out;
+    GF_ASSERT(this);
 
-    if (port) {
-        if (port > pmap->max_port)
-            goto out;
-    }
+    pmap = pmap_registry_get(this);
 
     if (brickname) {
-        p = pmap_registry_search(this, brickname, type, _gf_true);
+        p = pmap_registry_search(this, brickname, _gf_true);
         if (p)
             goto remove;
     }
 
     if (xprt) {
-        p = pmap_registry_search_by_xprt(this, xprt, type);
+        p = pmap_registry_search_by_xprt(this, xprt);
         if (p)
             goto remove;
     }
 
     goto out;
+
 remove:
     gf_msg("pmap", GF_LOG_INFO, 0, GD_MSG_BRICK_REMOVE,
            "removing brick %s on port %d", brickname, p);
 
-    if (xprt && (xprt == pmap->ports[p].xprt)) {
-        pmap->ports[p].xprt = NULL;
-    }
-
     /*
-     * This is where we garbage-collect.  If all of the brick names have
-     * been "whited out" by pmap_registry_search(...,destroy=_gf_true) and
-     * there's no xprt either, then we have nothing left worth saving and
-     * can delete the entire entry.
+     * If all of the brick names have been "whited out" by
+     * pmap_registry_search(...,destroy=_gf_true) and there's no xprt either,
+     *  then we have nothing left worth saving and can delete the entire entry.
      */
-    if (brick_disconnect || !pmap->ports[p].xprt) {
-        /* If the signout call is being triggered by brick disconnect
-         * then clean up all the bricks (in case of brick mux)
-         */
-        if (!brick_disconnect) {
-            brick_str = pmap->ports[p].brickname;
-            if (brick_str) {
-                while (*brick_str != '\0') {
-                    if (*(brick_str++) != ' ') {
-                        goto out;
+    cds_list_for_each_entry(tmp_port, &pmap->ports, port_list)
+    {
+        if (tmp_port->port == p) {
+            if (xprt && (tmp_port->xprt == xprt))
+                tmp_port->xprt = NULL;
+
+            if (brick_disconnect || !(tmp_port->xprt)) {
+                /* If the signout call is being triggered by brick disconnect
+                 * then clean up all the bricks (in case of brick mux)
+                 */
+                if (!brick_disconnect) {
+                    brick_str = tmp_port->brickname;
+                    if (brick_str) {
+                        while (*brick_str != '\0') {
+                            if (*(brick_str++) != ' ')
+                                goto out;
+                        }
                     }
                 }
+
+                tmp_port->brickname = NULL;
+                tmp_port->xprt = NULL;
+                ret = 0;
+                break;
             }
         }
-        free(pmap->ports[p].brickname);
-        pmap->ports[p].brickname = NULL;
-        pmap->ports[p].type = GF_PMAP_PORT_FREE;
+    }
+
+    if (!ret) {
+        cds_list_del_init(&tmp_port->port_list);
```

Comment:
We need to cleanup brickname also.

Reply:
If we are deleting the node from the list, won't that clear up its contents as well? Though, for the safer side, I am assigning the brickname and xprt to NULL in lines: 353, 354 before deleting the node. Shouldn't that do?

Reply:
if (!ret) {
        cds_list_del_init(&tmp_port->port_list);
        GF_FREE(tmp_port);
    }
Here you are cleanup memory for tmp_port but you are not cleanup brickname so it will become a leak.